extensions/websockets-next/deployment/src/test/java/io/quarkus/websockets/next/test/telemetry/OpenTelemetryWebSocketsTest.java

```diff
@@ -0,0 +1,202 @@
+
+    @TestHTTPResource("bounce")
+    URI bounceUri;
+
+    @TestHTTPResource("/")
+    URI baseUri;
+
+    @Inject
+    Vertx vertx;
+
+    @Inject
+    InMemorySpanExporter spanExporter;
+
+    @Inject
+    WebSocketConnector<BounceClient> connector;
+
+    @BeforeEach
+    public void resetSpans() {
+        spanExporter.reset();
+        BounceEndpoint.connectionId = null;
+        BounceEndpoint.endpointId = null;
+        BounceEndpoint.MESSAGES.clear();
+        BounceClient.MESSAGES.clear();
+        BounceClient.CLOSED_LATCH = new CountDownLatch(1);
+        BounceEndpoint.CLOSED_LATCH = new CountDownLatch(1);
+    }
+
+    @Test
+    public void testServerEndpointTracesOnly() {
+        assertEquals(0, spanExporter.getFinishedSpanItems().size());
+        try (WSClient client = new WSClient(vertx)) {
+            client.connect(new WebSocketConnectOptions(), bounceUri);
+            var response = client.sendAndAwaitReply("How U Livin'").toString();
+            assertEquals("How U Livin'", response);
+        }
+        waitForTracesToArrive(3);
+        assertServerTraces();
+    }
+
+    @Test
+    public void testClientAndServerEndpointTraces() throws InterruptedException {
+        var clientConn = connector.baseUri(baseUri).connectAndAwait();
+        clientConn.sendTextAndAwait("Make It Bun Dem");
+
+        // assert client and server called
+        Awaitility.await().untilAsserted(() -> {
+            assertEquals(1, BounceEndpoint.MESSAGES.size());
+            assertEquals("Make It Bun Dem", BounceEndpoint.MESSAGES.get(0));
+            assertEquals(1, BounceClient.MESSAGES.size());
+            assertEquals("Make It Bun Dem", BounceClient.MESSAGES.get(0));
+        });
+
+        clientConn.closeAndAwait();
+        // assert connection closed and client/server were notified
+        assertTrue(BounceClient.CLOSED_LATCH.await(5, TimeUnit.SECONDS));
+        assertTrue(BounceEndpoint.CLOSED_LATCH.await(5, TimeUnit.SECONDS));
+
+        waitForTracesToArrive(5);
+        assertServerTraces();
+        assertClientTraces();
+    }
+
+    @Test
+    public void testServerTracesWhenErrorOnMessage() {
+        assertEquals(0, spanExporter.getFinishedSpanItems().size());
+        try (WSClient client = new WSClient(vertx)) {
+            client.connect(new WebSocketConnectOptions(), bounceUri);
+            var response = client.sendAndAwaitReply("It's Alright, Ma").toString();
+            assertEquals("It's Alright, Ma", response);
+            response = client.sendAndAwaitReply("I'm Only Bleeding").toString();
+            assertEquals("I'm Only Bleeding", response);
+
+            client.sendAndAwait("throw-exception");
+            Awaitility.await().atMost(Duration.ofSeconds(5)).until(client::isClosed);
+            assertEquals(WebSocketCloseStatus.INTERNAL_SERVER_ERROR.code(), client.closeStatusCode());
+        }
+        waitForTracesToArrive(3);
+        assertServerTraces();
```

Comment:
I dropped the assertion methods. Regarding testing of recorder exceptions - I am not quite sure how to do that, especially in case of connection on open I don't know how to prepare such a state. The method that tests error on message just checks that both open/close are still recorder in case of an error.

Reply:
I guess you would need to set a very low timeout and hold the connection on the other side or, try to connect to a non-existing endpoint. The hard part is tune the timeout because the test has to be fast enough.